### PR TITLE
init object for snapsync pipeline

### PR DESCRIPTION
**Pull request by matkt**
Signed-off-by: Karim TAAM <karim.t2am@gmail.com>

## PR description

This is the first step initializing the snapsync pipeline. Setting up utility objects and request classes. 

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StackTrie.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    this.keys.putAll(keys);
+  }
+
+  public void addProofs(final List<Bytes> proofs) {
+    this.proofs.addAll(proofs);
+  }
+
+  public void commit(final NodeUpdater nodeUpdater) {
+    if (nbSegments.decrementAndGet() <= 0 && (!proofs.isEmpty() || !keys.isEmpty())) {
+      final Map<Bytes32, Bytes> proofsEntries = Collections.synchronizedMap(new HashMap<>());
```
Comment (fab-10) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StackTrie.java:71]: is a synchronized map required here? it is only read in other places.
This applies also to other requests in this package

Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StackTrie.java:71]: nop this is a bad copy past

File: ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/Node.java [hunk 2]
```diff
@@ -71,4 +71,16 @@ default boolean isReferencedByHash() {
...
+   * Return if a node needs heal. If one of its children missing in the storage
+   *
+   * @return true if the node need heal
+   */
+  boolean isNeedHeal();
+
+  /**
+   * Marking a node as need heal means that one of its children is not yet present in the storage
+   */
+  void markNeedHeal();
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/Node.java:85]: done

File: ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/Node.java [hunk 3]
```diff
@@ -71,4 +71,16 @@ default boolean isReferencedByHash() {
 
   /** Unloads the node if it is, for example, a StoredNode. */
   default void unload() {}
+
+  /**
+   * Return if a node needs heal. If one of its children missing in the storage
+   *
+   * @return true if the node need heal
+   */
+  boolean isNeedHeal();
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/Node.java:80]: done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldDownloadState.java [hunk 4]
```diff
@@ -119,6 +121,13 @@ private synchronized void cleanup(final Void result, final Throwable error) {
     }
   }
 
+  protected void cleanupQueues() {
+    for (final EthTask<?> outstandingRequest : outstandingRequests) {
+      outstandingRequest.cancel();
+    }
+    pendingRequests.clear();
+  }
+
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldDownloadState.java:130]: this method is call in another synchronized method. So it's fine. But I can add a synchronized if someone want to  onlycall this method in the future

[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldDownloadState.java:130]: done

File: ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/SnapPutVisitor.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.trie;
+
+import org.apache.tuweni.bytes.Bytes;
+
+public class SnapPutVisitor<V> extends PutVisitor<V> {
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/SnapPutVisitor.java:19]: Yes I will try to add that

[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/SnapPutVisitor.java:19]: but it is already tested indirectly with the tests of the stacktrie

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StackTrie.java (hunk 1 above)
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StackTrie.java:71]: I hope it will be the last one I will miss

**Review by fab-10 — APPROVED** [review]

**Pull request merged**